[WebXR] Add WebXR backend to support immersive-ar backed by ARKit

djg opened the pull request:
#### 7ad679f071434da45d453251ba7446294a54f57d
<pre>
[WebXR] Add WebXR backend to support immersive-ar backed by ARKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=262679">https://bugs.webkit.org/show_bug.cgi?id=262679</a>
rdar://116502210

Reviewed by Dean Jackson.

ARKitCoordinator adds just enough support for WebXR page to query support for
immersive-ar, allowing the XR immersive-ar-session.html button to show `START
AR` instead of `AR NOT FOUND`

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Shared/XR/XRDeviceInfo.cpp:
(WebKit::XRDeviceInfo::encode const):
(WebKit::XRDeviceInfo::decode):
* Source/WebKit/Shared/XR/XRDeviceInfo.h:
* Source/WebKit/Shared/XR/XRDeviceProxy.cpp:
(WebKit::XRDeviceProxy::XRDeviceProxy):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.h: Added.
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm: Added.
(WebKit::ARKitCoordinator::getPrimaryDeviceInfo):
(WebKit::ARKitCoordinator::requestPermissionOnSessionFeatures):
(WebKit::ARKitCoordinator::startSession):
(WebKit::ARKitCoordinator::endSessionIfExists):
(WebKit::ARKitCoordinator::scheduleAnimationFrame):
(WebKit::ARKitCoordinator::submitFrame):
(WebKit::ARKitCoordinator::connectSessionIfNeeded):
* Source/WebKit/UIProcess/XR/ios/PlatformXRSystemIOS.mm: Copied from Source/WebKit/Shared/XR/XRDeviceInfo.h.
(WebKit::PlatformXRSystem::xrCoordinator):
* Source/WebKit/UIProcess/ios/WebProcessProxyIOS.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/269049@main">https://commits.webkit.org/269049@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a841fa3673ea3c8cb4f868e061765dd90a0eb30

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21404 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21712 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22457 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23266 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19836 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21645 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/25013 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21959 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21034 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21629 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21272 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18545 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24118 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18443 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19394 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25715 "Passed tests") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/18615 "Found 1 new JSC binary failure: testapi (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19528 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19602 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23563 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/20784 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20087 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17101 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/cache-storage/worker/cache-add.https.html (failure)") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/24809 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19405 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5864 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5121 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23645 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/26079 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19993 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5691 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

grorg commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1618,3 +1618,7 @@
 #define HAVE_AUDIOFORMATPROPERTY_VARIABLEPACKET_SUPPORTED 1
 #endif
+
+#if PLATFORM(IOS) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 110000
```
I don't think we need the `MIN_REQUIRED` since we don't support that far back any more.

djg committed:
[WebXR] Add WebXR backend to support immersive-ar backed by ARKit
https://bugs.webkit.org/show_bug.cgi?id=262679
rdar://116502210

Reviewed by Dean Jackson.

ARKitCoordinator adds just enough support for WebXR page to query support for
immersive-ar, allowing the XR immersive-ar-session.html button to show `START
AR` instead of `AR NOT FOUND`

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Shared/XR/XRDeviceInfo.cpp:
(WebKit::XRDeviceInfo::encode const):
(WebKit::XRDeviceInfo::decode):
* Source/WebKit/Shared/XR/XRDeviceInfo.h:
* Source/WebKit/Shared/XR/XRDeviceProxy.cpp:
(WebKit::XRDeviceProxy::XRDeviceProxy):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.h: Added.
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm: Added.
(WebKit::ARKitCoordinator::getPrimaryDeviceInfo):
(WebKit::ARKitCoordinator::requestPermissionOnSessionFeatures):
(WebKit::ARKitCoordinator::startSession):
(WebKit::ARKitCoordinator::endSessionIfExists):
(WebKit::ARKitCoordinator::scheduleAnimationFrame):
(WebKit::ARKitCoordinator::submitFrame):
(WebKit::ARKitCoordinator::connectSessionIfNeeded):
* Source/WebKit/UIProcess/XR/ios/PlatformXRSystemIOS.mm: Copied from Source/WebKit/Shared/XR/XRDeviceInfo.h.
(WebKit::PlatformXRSystem::xrCoordinator):
* Source/WebKit/UIProcess/ios/WebProcessProxyIOS.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/269049@main
